File: Problem106.cpp
//106. Construct Binary Tree from Inorder and Postorder Traversal - M. Given inorder and postorder traversal of a tree, construct the binary tree.
//Note: You may assume that duplicates do not exist in the tree.
//IMPORTANT FUNCTION: std::distance(vecOfNums.begin(), it);

#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;


// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

TreeNode* buildTree(vector<int>& inorder, int inStart, int inEnd,
                          vector<int>& postorder, int postStart, int postEnd) {
    if (inStart > inEnd || postStart > postEnd)
        return NULL;
    int rootValue = postorder[postEnd];
    TreeNode* root = new TreeNode(rootValue);

    int k = 0;
    for (int i = 0; i < inorder.size(); i++) {
        if (inorder[i] == rootValue) {
            k = i;
            break;
        }
    }

    root->left = buildTree(inorder, inStart, k - 1, postorder, postStart, postStart + k - (inStart + 1));
    root->right = buildTree(inorder, k + 1, inEnd, postorder, postStart + k- inStart, postEnd - 1);

    return root;
}

TreeNode* buildTree(vector<int>& inorder, vector<int>& postorder) {
    auto res = buildTree(inorder,0 , inorder.size()-1, postorder, 0, postorder.size()-1);
    return res;
}


int main() {
    vector<int> inorder {9,3,15,20,7};
    vector<int> postorder {9,15,7,20,3};

    TreeNode* tn = buildTree(inorder, postorder);
    delete(tn);

    return 0;
}

File: NewJob2025/Problem106.java
// 106. Construct Binary Tree from Inorder and Postorder Traversal

class Solution {
    public int findPos(int[] inorder, int inL, int inR, int val) {
        for (int i = inL; i <= inR; i++) {
            if (inorder[i] == val) return i;
        }

        return -1;
    }

    public TreeNode getRoot(int[] inorder, int inL, int inR, int[] postorder, int postL, int postR) {
        TreeNode currRoot = new TreeNode(postorder[postR]);
        if (postL == postR) return currRoot;

        int rootPos = findPos(inorder, inL, inR, postorder[postR]);

        if (rootPos > inL) {
            currRoot.left = getRoot(inorder, inL, rootPos - 1, postorder, postL, postL + (rootPos - inL - 1));
        }

        if (rootPos < inR) {
            currRoot.right = getRoot(inorder, rootPos + 1, inR, postorder, postR - (inR-rootPos), postR - 1);
        }

        return currRoot;
        
    }

    public TreeNode buildTree(int[] inorder, int[] postorder) {
        return getRoot(inorder, 0, inorder.length - 1, postorder, 0, postorder.length - 1);
    }
}
